### PR TITLE
[Draft] use Login for the top right corner

### PR DESCRIPTION
**Pull request by danielma**
# The Big Idea

We have a long-term goal of moving account related items to the Login app. We already replaced the logout link, and this PR _also_ moves a) unlinking accounts, b) switching accounts, and c) loading the list of connected accounts.

I'm doing this work now, at the behest of an H1 report about unlinking accounts with `GET` requests. It's no critical security issue, but we don't want customer data _changing_ in `GET` requests.

# Guiding Principles

Quoting from [the honorable Topbar Story](https://github.com/planningcenter/design/blob/master/planningcenter/topbar/docs/STORY.md#expectations):

> 1. And you shall love the API with all your heart, soul and mind.
> 2. The second is like it: work on any framework as you work on Rails.

These mutating actions (unlink and switch) _used_ to happen with session authenticated `GET` requests to a Rails app. Moving to a `POST` isn't immediately straightforward, because our apps require CSRF tokens to verify `POST` requests. An easy answer might have been to follow the pattern of `PlatformAnnouncements`, get the CSRF token from a `<meta>` tag, and move on. My reading of the Story (and talks with @chantastic), however, left me feeling like this wasn't the right path for Topbar.

Requiring a `<meta>` tag on the host page, and/or adding a new `csrfToken` prop to exported components both fell short of the desire to work on any framework as you work on Rails.

And so, the path forward in this PR is to `POST` to the API, and then `window.location` to the `meta.redirect_to`. It gets us the `POST` we desire, and the continued possibility of being mounted outside a Rails app.

# New Host Application requirements

`AppsProvider`, `ConnectedPeopleProvider`, and `PlatformAnnouncements` now take an `authenticatedFetch` prop, which is a function with an identical signature to `window.fetch`, and must be able to make authenticated requests (of any verb) to the PCO API.

The default fetch function built into the project will successfully make `GET` requests when mounted in a Rails app using session authentication, but will fail on action verbs because of the need for a CSRF token. [This draft PR into Giving](https://github.com/ministrycentered/giving/pull/4482) shows how a host app will probably need to adjust to support this new version of Topbar.

Giving adds the CSRF header, and `{ credentials: "include", mode: "cors" }`. If, for example, an app needs to support a browser without `fetch` and doesn't have a polyfill, that app can inject any function that conforms to the protocol and makes an authenticated request.

# Final Thoughts

In accordance with The Story, I feel good about these proposed changes and the new requirements for host apps. Topbar has always been designed to talk to the API, and clarifying the responsibility of the host app to provide a way to talk to the API seems to be a good responsibility line for Topbar. It also removed a Topbar dependency, which felt great.

I recognize that by opening this PR, I'm stepping into a repo I've never contributed to. Please take all my thoughts and changes with humble grain of salt. I am open and ready for all feedback that may come 🙂

### CODE REVIEW

**Comment by chantastic** [comment]
@danielma thanks for opening this PR!
just wanted to let you know that I see it, might not be able to discuss it today.

**Comment by danielma** [comment]
@chantastic I've updated with a real description!

**Comment by chantastic** [comment]
@danielma your description was an extremely enjoyable read.

thank you for very diligently considering the history and design intent into proposed solution!

you know i'm a sucker for injectables and removing dependencies.

i would say, **lean into the standard `fetch` return type**

`superagent` was added as a Platform acceptance criteria when `fetch` was not widely supported and and apps didn't universally support `jQuery.getJSON`.  
there was some concerns about the universal ability of apps to polyfill `fetch`.  
so superagent was selected as a stop-gap.

with our current browser support, we really don't *have* to polyfill `fetch`.
doing so would just be a courtesy.

**are there other developers that should have a chance to review and discuss this PR?**

**Comment by chantastic** [comment]
@danielma crossing the new week boundary, just wanted to ping you and see if you want wanted to move forward with the `fetch` type and solicit any further review from other devs

**Comment by danielma** [comment]
@chantastic thanks for the ping! Sorry I was slow to respond. I was out on Friday and have had a full schedule with reviews.

I will definitely make changes to push the API towards `fetch`, and I will grab a few developers to vet these changes as well, but it might be a few days before I get to it

**Comment by chantastic** [comment]
thanks @danielma!
to clarify, no rush on my end.
i just try to check in on live ones at the beginning of the week :)

**Comment by danielma** [comment]
@chantastic I've updated this PR with a new interface for authenticated requests to the API. Now, rather than conforming to the arbitrary `ApiRequest` type I had defined, we ask for a function that conforms to the standard `fetch` signature.

File: planningcenter/topbar/modules/api_request.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+export type Fetch = typeof fetch;
+
+export const defaultFetch: Fetch = (url, options) => {
+  return fetch(url, { credentials: "include", mode: "cors", ...options });
+};
```
Comment (danott) [inline planningcenter/topbar/modules/api_request.ts:5]: I understand the sentiment of "the first and second commandments of TopBar". The aspirations of being able to work with any framework as we would with Rails feels great in the abstract... but it has yet to become a reality.

I wonder if there's a middle ground here? My idea would be that the `defaultFetch` can provide some out of the box affordance for Rails sessions, while remaining open to being overridden in the hypothetical future of a non-Rails application that uses some other method of API authentication.

```suggestion
export const defaultFetch: Fetch = (url, options) => {
  const railsCsrfToken: null | string = inferRailsCsrfToken()
  if (railsCsrfToken === null) {
    return fetch(url, { credentials: "include", mode: "cors", ...options });
  } else {
    return railsCsrfFetch(url, pseuedoCodeBuildUpTheOptions);
  }
};

// copied from https://github.com/ministrycentered/giving/pull/4482
function railsCsrfFetch(url, { headers, ...rest) {
   return fetch(url, {
     headers: { ...headers, "X-CSRF-Token": inferRailsCsrfToken() },
     ...rest,
     credentials: "include",
     mode: "cors",
   })
}

function inferRailsCsrfToken(): null | string {
  return window.Rails?.csrfToken?() || document.querySelector('meta[name="csrf-token"]')?.content || null
}
```

Seeing `credentials: "include"` and `mode: "core"` in both branches of this idea, it _seems_ like we already require some knowledge from Top Bar for what our infrastructure expects from a fetch request?

File: planningcenter/topbar/modules/apps_provider.tsx [hunk 2]
```diff
@@ -1,11 +1,12 @@
 import * as React from "react";
 import pcoUrl from "./pco_url";
-import getJSON from "./get_json";
+import apiRequest, { Fetch, defaultFetch } from "./api_request";
 
 export class AppsProvider extends React.Component<
   {
     env: string;
     formatter?: any;
+    authenticatedFetch?: Fetch;
```
Comment (danott) [inline planningcenter/topbar/modules/apps_provider.tsx:9]: Authentication is the current goal of the consumer defining their own `fetch`. I wonder if that will always be the case? I find myself wanting to call this `configuredFetch`, or maybe even `overridenFetch`, to open up the possibilities of doing _whatever_ you want to do by intercepting the `fetch` arguments before it's in flight.

**Review by danott — COMMENTED** [review]
The proposed approach is clear in my eyes. As noted in my one giant comment, I think it'd be worth considering some batteries-included support for our Rails sessions, since every consumer of the top bar is currently in a Rails application, that uses session authentication to access the API.

The idea of every application defining a function nearly identical to what is defined in Giving's draft Pull Request feels a little bit...icky?

A process aside: I love TypeScript.

File: planningcenter/topbar/modules/api_request.ts (hunk 1 above)
Comment (danielma) [reply planningcenter/topbar/modules/api_request.ts:5]: You make a good point here. I wanted to stay blissfully unaware of Rails, but that's a pretty abstract goal.

FWIW on `credentials` and `mode`, my thinking in including them by default is that they would work in any host application that used session authentication to talk to the API. The CSRF token implementation detail is Rails specific, and that's why I stopped short of it.

@chantastic I'd love to hear your thoughts on this. If you feel ok about adding the CSRF inference, I do too.

File: planningcenter/topbar/modules/apps_provider.tsx (hunk 2 above)
Comment (danielma) [reply planningcenter/topbar/modules/apps_provider.tsx:9]: `configuredFetch` seems like a great option here. It leaves that door open to anything else the host app might need to do to make a good request. Thanks for that suggestion

**Review by sheck — COMMENTED** [review]
This looks cool! I agree with @danott's comments though — if we won't be using topbar on any non-rails app yet, it does feel strange to require all of the rails apps to provide a nearly identical function. I'm a fan of the approach of providing a good rails default that can be overridden (so we don't preclude support for potential non-rails apps in the future).

File: planningcenter/topbar/modules/api_request.ts (hunk 1 above)
Comment (chantastic) [reply planningcenter/topbar/modules/api_request.ts:5]: i think including a Rails-specific fetch function still aligns with the goals, provided that it's still injectable.

suffice to say, i'm not dogmatic about not having Rails specific implementations exported — as long as they aren't pre-baked into the implementation.

If it were me, I'd forgo `defaultFetch` and export only `railsCsrfFetch` and `inferRailsCsrfToken`.
Folks can use these by their explicit names or write a custom function product side.

I like the naming orientation of `railsCsrfFetch` as you could naturally have a `nodeCsrfFetch` and `universalCsrfFetch` — should those days ever come.

File: planningcenter/topbar/modules/apps_provider.tsx (hunk 2 above)
Comment (chantastic) [reply planningcenter/topbar/modules/apps_provider.tsx:9]: i too prefer `configuredFetch` to `overridenFetch`.
i think the connotation is better — more closely aligned to the actual pattern: injected fetch

File: planningcenter/topbar/modules/api_request.ts (hunk 1 above)
Comment (danielma) [reply planningcenter/topbar/modules/api_request.ts]: @chantastic the latest revision doesn't export anything from the `api_request` module. It retains only the `configuredFetch` API. If there's an app that needs to do something special to configure their fetch, I don't expect that app will gain anything from Topbar exporting functions to infer CSRF tokens. 

`api_request.ts` exports `defaultFetch` only to other modules within Topbar, not as part of the final bundle.

File: planningcenter/topbar/modules/apps_provider.tsx (hunk 2 above)
Comment (danielma) [reply planningcenter/topbar/modules/apps_provider.tsx]: `configuredFetch` is in the latest change

File: planningcenter/topbar/modules/api_request.ts (hunk 1 above)
Comment (chantastic) [reply planningcenter/topbar/modules/api_request.ts]: @danielma i gotcha. i'm mostly curious why we're abstracting over `railsCsrfFetch` with `defaultFetch`.

if the intent is to buy into a rails environment by default, my preference would be to see direct communication of that default in the naming.

`defaultFetch` feels like name we give the generic (non-CSRF) version, *after* this is a multi-platform solution.

I think I see two possible outcomes:

**1. Keep it pure** 
* use a `defaultFetch` (without CSRF) by default
* export (as part of the public API) `railsCsrfFetch`
* require composition app-side

this is probably the most precedented pattern and virtually eliminates a need for a breaking defaults change down the line

**2. Screw purity. We live in Rails**
* make the default Rails-dependent by default
* don't export anything. don't require app-side composition
* what and see if we ever need anything else

of course, there's a part of me that feels like — even when we can guarantee Rails — it's appropriate to provide the opt-out `fetch` with the default options you've included.

---

Question: is there a practical environment that we have today that could need a non-CSRF version for texting and what not.

In almost all cases, I feel like we've made it easy enough by including the Rails-ready function. Exporting it, and requiring app-side composition, doesn't feel like over-reaching. And it has the benefit of keeping the library logically decoupled while being Rails ready.

Just some thoughts.

**just incase i'm still being too opaque**
my inclination would be to export `railsCsrfFetch` as public Topbar API.
we'd remove the conditional and rely on non-CSRF logic branch as a default.
apps would then import and inject `railsCsrfFetch` app-side.

I'm open to that being a step too far.
However, I've seen app-side injection save us SO MANY TIMES.

But I'm also open to Option 2, shipping it, and peeling out the platform-specific default when needed.

Comment (chantastic) [reply planningcenter/topbar/modules/api_request.ts]: i'm sorry if that's only 60% cogent.
trying to wrap as many loose ends as i can before being gone 'til wednesday.

Comment (danielma) [reply planningcenter/topbar/modules/api_request.ts:13]: @chantastic thanks for all the clarification. I'm sorry I didn't understand it the first time. With this new clarity, I can see you were saying the same thing in your last comment, and I just missed it. 

I'm on board with this design, as a mid-point between the initial version (require apps to provide their own Rails-friendly fetch method) and the secondary (include Rails-specific behavior by default).

**Comment by chantastic** [comment]
thanks for taking time to comb thru it!

looks good to me. i feel good about merging some kind of pre-major release.
i'm out today but could get that done if there are no remaining items.

**Comment by chantastic** [comment]
scratch that, i'm gonna merge it no and you can continue to explore...

**Comment by chantastic** [comment]
catch that again, i can't until it's out of draft.

i'll just take care of it later in the day.
would you mind un-drafting it tho so we can merge? thanks :)

**Comment by chantastic** [comment]
@danielma thank you for your dedication to this project and process! i'm grateful for all your hard work.

i've published version [7.0.0-alpha.0](https://www.npmjs.com/package/@planningcenter/topbar/v/7.0.0-alpha.0) to npm with your changes.
please start exploring it in the apps you have most control over and — if all goes well — we can start working up a migration doc when I get back (Wednesday).

**Pull request merged**